Remove duplicate is-clause? fn

appleby opened the pull request:
Closes #39333

### Description

Move `mbql-clause?`, `is-clause?`, and `check-clause` from `mbql.u` to `schema.helpers`, and remove the duplicate definition of `is-clause?`.

If we want to reduce churn, we could just have `mbql.u` re-export the symbols from `schema.helpers`, but it seemed better to update all callers to `require` them from the new location.

h/t to @sdduursma as this is a re-implementation of his changes in #40945

### How to verify

No new tests added, but these funcs should be well covered by existing tests.

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

appleby committed:
Move mbql.u/{mbql-clause?,is-clause?,check-clause} to schema.helpers

And remove duplicate definitions.

Closes #39333

bshepherdson reviewed (changes requested):
I actually think we do want to re-export from `mbql.u`. That's quite common in our `util` namespaces and I find it surprising to be accessing `foo.schema.helpers` from outside `foo.schema` and even more so outside `foo.*`.

appleby commented:
> I actually think we do want to re-export from `mbql.u`. That's quite common in our `util` namespaces and I find it surprising to be accessing `foo.schema.helpers` from outside `foo.schema` and even more so outside `foo.*`.

Yeah fair enough, was on the fence about this. Will update.

appleby commented:
> > I actually think we do want to re-export from `mbql.u`. That's quite common in our `util` namespaces and I find it surprising to be accessing `foo.schema.helpers` from outside `foo.schema` and even more so outside `foo.*`.
> 
> Yeah fair enough, was on the fence about this. Will update.

Done. Fwiw, there are a couple of existing places where we were already using `schema.helpers` outside of the `legacy-mbql` namespace, but they are the exception rather than the rule.

bshepherdson commented on src/metabase/legacy_mbql/schema/helpers.cljc:
```diff
@@ -9,6 +9,36 @@
 
+(defn mbql-clause?
+  "True if `x` is an MBQL clause (a sequence with a keyword as its first arg). (Since this is used by the code in
+  `normalize` this handles pre-normalized clauses as well.)"
```
Good catch... I'm not sure. In practice there's no harm in letting this function be more flexible.

appleby replied:
I guess I picked the wrong example func. Possibly the point of the versions in the normalize ns is to avoid constantly and unnecessarily calling `maybe-normalize-token` in `is-clause?` for something that should already be normalized. I'll leave the local versions in normalize alone and just update the docstring here.

```clojure
(defn is-clause?
  "If `x` an MBQL clause, and an instance of clauses defined by keyword(s) `k-or-ks`?

    (is-clause? :count [:count 10])        ; -> true
    (is-clause? #{:+ :- :* :/} [:+ 10 20]) ; -> true

  (This is different from the implementation in `mbql.u` because it also supports un-normalized clauses. You shouldn't
  need to use this outside of this namespace.)"
  [k-or-ks x]
  (and
   (mbql-clause? x)
   (let [clause-name (maybe-normalize-token (first x))]
     (if (coll? k-or-ks)
       ((set k-or-ks) clause-name)
       (= k-or-ks clause-name)))))
```

appleby committed:
Fix docstring of schema.helpers.mbql-clause?

This func is apparently no longer used by code in `normalize`, which has it's own local def.
